File: othello_python/othello.py
#!/usr/bin/env python3
"""
Class Othello + main
Play Othello (Reversi) on the command line
https://en.wikipedia.org/wiki/Reversi
Akseli Lukkarila
2019
"""
import sys

from board import Board
from colorprint import Color, print_bold, print_error
from player import Player
from utils import Disk, clamp


class Othello:
    """Play Othello CLI game."""

    def __init__(self, board_size: int):
        self.board = Board(board_size)
        self.player_black = Player(Disk.BLACK)
        self.player_white = Player(Disk.WHITE)
        self.rounds_played = 0
        self.games_played = 0
        self.board_size = board_size

    def init_game(self):
        """Initialize game board and players for a new game."""
        if self.games_played > 0:
            self.board = Board(self.board_size)
            self.player_black.reset()
            self.player_white.reset()
            self.rounds_played = 0

        if self.get_answer("Would you like to play against the computer"):
            if self.get_answer("Would you like to play as black or white", yes="b", no="w"):
                self.player_white.set_human(False)
            else:
                self.player_black.set_human(False)

        print_bold("\nPlayers:")
        self.print_status()

    def play(self):
        """Play one full game of Othello."""
        self.init_game()
        self.game_loop()
        self.print_result()
        if self.get_answer("\nWould you like to play again"):
            self.play()

    def game_loop(self):
        """Keep making moves until both players can't make a move anymore."""
        while self.board.can_play() and (self.player_black.can_play() or self.player_white.can_play()):
            self.rounds_played += 1
            print_bold(f"\n=========== ROUND: {self.rounds_played} ===========")
            self.player_black.play_one_move(self.board)
            print("-------------------------------")
            self.player_white.play_one_move(self.board)

        self.games_played += 1

    def print_result(self):
        """Print ending status and winner info."""
        print_bold("\n===============================")
        print_bold("The game is finished!\n", Color.green)
        print_bold("Result:")
        self.print_status()
        print("")

        winner = self.board.result()
        if winner == Disk.EMPTY:
            print_bold("The game ended in a tie...")
        else:
            print_bold(f"The winner is {winner}!")

    def print_status(self):
        """Print current board and player info."""
        print(self.player_black)
        print(self.player_white, end="\n\n")
        print(self.board)

    @staticmethod
    def get_answer(question: str, yes="y", no="n") -> bool:
        """Ask a question with two options, and return bool from user answer."""
        ans = input(f"{question} ({yes}/{no})? ")
        return ans.lower() == yes.lower()

    @staticmethod
    def get_board_size() -> int:
        """Ask and return the desired board size."""
        while True:
            try:
                ans = int(input("Choose board board_size (default is 8): "))
                return clamp(ans, 4, 16)
            except ValueError:
                print_error("give a number...")


if __name__ == "__main__":
    print_bold("OTHELLO GAME - Python", Color.green)
    args = sys.argv[1:]
    try:
        # try to read board size from command line args
        board_size = int(args[0])
        if board_size < 4 or board_size > 16:
            print_error("Unsupported board size: {}", board_size)
            sys.exit(1)
        print(f"Using board size: {board_size}")
    except (ValueError, IndexError):
        board_size = Othello.get_board_size()

    try:
        game = Othello(board_size)
        game.play()
    except KeyboardInterrupt:
        sys.exit("\nGame aborted...")

File: othello_kotlin/app/src/main/kotlin/othello/Othello.kt
package othello

import java.awt.Color
import kotlin.system.exitProcess

/** Gameplay loop and main logic.*/
class Othello(private val settings: Settings) {
    private var board = Board(settings.boardSize)
    private val playerBlack = Player.black(settings.toPlayerSettings())
    private val playerWhite = Player.white(settings.toPlayerSettings())
    private var roundsPlayed = 0
    private var gamesPlayed = 0
    private var gameLog = mutableListOf<String>()

    /** Play one full game of Othello.*/
    fun play() {
        while (true) {
            initGame()
            gameLoop()
            printResult()
            if (settings.showLog) {
                printLog()
            }
            if (settings.autoplayMode || !getAnswer("Would you like to play again")) {
                break
            }
        }
    }

    /** Initialize game board and players for a new game.*/
    private fun initGame() {
        if (gamesPlayed > 0) {
            board = Board(settings.boardSize)
            playerBlack.reset()
            playerWhite.reset()
            roundsPlayed = 0
            gameLog.clear()
        }

        if (settings.autoplayMode) {
            // Computer plays both
            playerBlack.setHuman(false)
            playerWhite.setHuman(false)
        } else if (settings.quickStart) {
            // Default: play as black against white computer player
            playerWhite.setHuman(false)
        } else if (getAnswer("Would you like to play against the computer")) {
            if (getAnswer("Would you like to play as black or white", "b", "w")) {
                playerWhite.setHuman(false)
            } else {
                playerBlack.setHuman(false)
            }
        }

        printBold("\nPlayers:")
        printStatus()
    }

    /** Keep making moves until both players can't make a move any more.*/
    private fun gameLoop() {
        while (board.canPlay() && (playerBlack.canPlay || playerWhite.canPlay)) {
            roundsPlayed++
            printBold("\n=========== ROUND: $roundsPlayed ===========")
            for (player in listOf(playerBlack, playerWhite)) {
                val result = player.playOneMove(board)
                if (result != null) {
                    gameLog.add("$result;${board.toLogEntry()}")
                }
            }
            println("--------------------------------")
        }
        gamesPlayed++
    }

    private fun printLog() {
        val formattedLog = gameLog
            .mapIndexed { index, line -> String.format("%02d: %s", index + 1, line) }
            .joinToString("\n")

        val hexHash = calculateSha256(formattedLog)

        printBold("Game log:", Color.YELLOW)
        println(formattedLog)
        println(hexHash)
    }

    /** Print ending status and winner info.*/
    private fun printResult() {
        printBold("\n================================")
        printColor("The game is finished!", Color.GREEN)
        printBold("Result:")
        printStatus()
        println()

        val winner = board.result()
        if (winner == Disk.Empty) {
            printBold("The game ended in a tie...\n")
        } else {
            printBold("The winner is ${winner.name()}!\n")
        }
    }

    /** Print current board and player info.*/
    private fun printStatus() {
        println(playerBlack)
        println(playerWhite)
        println("\n$board")
    }

    /** Ask a question with two options, and return bool from user answer.*/
    private fun getAnswer(question: String, yes: String = "y", no: String = "n"): Boolean {
        print("$question ($yes/$no)? ")
        val ans = readlnOrNull()
        return !ans.isNullOrBlank() && ans.equals(yes, ignoreCase = true)
    }
}

fun main(args: Array<String>) {
    printBold("OTHELLO GAME - KOTLIN", Color.GREEN)

    if (args.contains("--help") || args.contains("-h")) {
        println("Othello Kotlin ${versionInfo()}\n")
        println("USAGE: java -jar othello_kotlin.jar [board size]\n")
        println("Optional arguments:")
        println("    -h | --help          Print usage and exit")
        println("    -v | --version       Print version info and exit")
        exitProcess(1)
    }

    if (args.contains("--version") || args.contains("-v")) {
        println("Othello Kotlin ${versionInfo()}")
        exitProcess(0)
    }

    // Try to read board size from command line args
    val boardSize = if (args.isEmpty() || args[0].toIntOrNull() == null) {
        getBoardSize()
    } else {
        val size = args[0].toInt()
        if (size < MIN_BOARD_SIZE || size > MAX_BOARD_SIZE) {
            printError("Unsupported board size: $size")
        }
        println("Using board size: $size")
        size
    }

    val settings = Settings(
        boardSize,
        autoplayMode = true,
        quickStart = false,
        showHelpers = true,
        showLog = true,
        testMode = true,
    )

    val game = Othello(settings)
    game.play()
}

/** Ask and return the desired board size.*/
private fun getBoardSize(): Int {
    print("Choose board size (default is $DEFAULT_BOARD_SIZE): ")
    val input = readlnOrNull()
    if (input != null) {
        val boardSize = input.toIntOrNull()
        if (boardSize != null) {
            if (boardSize !in MIN_BOARD_SIZE..MAX_BOARD_SIZE) {
                printWarn("Limiting board size to valid range $MIN_BOARD_SIZE...$MAX_BOARD_SIZE")
            }
            return boardSize.coerceIn(MIN_BOARD_SIZE, MAX_BOARD_SIZE)
        }
    }
    printWarn("Invalid size, defaulting to $DEFAULT_BOARD_SIZE...")
    return DEFAULT_BOARD_SIZE
}
